common/app/Map/Map.jsx

```diff
@@ -6,18 +6,76 @@ import { Col, Row } from 'react-bootstrap';
 import ns from './ns.json';
 import SuperBlock from './Super-Block.jsx';
 import { superBlocksSelector } from '../redux';
+import { initialNodeSelector } from './redux';
 
 const mapStateToProps = state => ({
-  superBlocks: superBlocksSelector(state)
+  superBlocks: superBlocksSelector(state),
+  initialNode: initialNodeSelector(state)
 });
 
 const mapDispatchToProps = {};
 const propTypes = {
+  initialNode: PropTypes.string,
   params: PropTypes.object,
   superBlocks: PropTypes.array
 };
 
 export class ShowMap extends PureComponent {
+  componentDidUpdate({ initialNode: prevInitialNode }) {
+    const { initialNode } = this.props;
+    if ( initialNode !== prevInitialNode ) {
+      this.updateMapScrollAttempts = 0;
+      this.updateMapScroll();
+    }
+  }
+
+  updateMapScroll() {
+    const { initialNode } = this.props;
+    const rowNode = this._row;
+    const challengeNode = rowNode.querySelector(
+      `[data-challenge="${initialNode}"]`
+    );
+
+    if ( !challengeNode ) {
```

Comment:
Why would challengeNode be undefined?

Reply:
@BerkeleyTrue in case it's not in the DOM yet.

Reply:
I see. In that case I suggest we remove the retry logic and the set timeout logic and trigger this method on CDM. This should prevent trying to update dom nodes on the server/preCDM

Reply:
CDM is called too long before challenges are fetched. That is why CDU is used, and update is triggered only when initialNode prop is changed (on challenge.fetch.complete).
So timeout from CDM is hard to predict. However timeout from CDU is more predictable, but is less reliable than retry IMO.

Reply:
> CDM is called too long before challenges are fetched.

In my scenario we would keep both triggers. The SCU when the dom is live and the CDM to be the initial trigger. That way if SCU happens when the element isn't mounted the CDM catches it. This avoids the need for the retry logic.

> I didn't get that part as I'm not aware of the ssr part.

The way it's currently set up, you'll be setting up a timeout for every request on the server. Since the dom will never mount server side, every request will create 5 setTimeouts, potentially creating side effects after a request has been fulfilled. You usually want to avoid logic like this, and with the hooks that react provides it is usually quite trivially to do so.

Reply:
To be honest I'm completely confused right now. I've set loggers to CDM and CDU methods of Map. None of them is called on the server during the ssr, only render is called once.
Am I doing something wrong? When is CDM called on server?

Anyway, I'm ok to shift from retry logic to single timeout. In my first attempts 300ms on CDU was enough to properly setup scroll on Map. What do you think?

Reply:
Your right, both of those are not called on the sever. So the server side argument is invalid.

I've pulled down your and I'm finding it hard to replicate the missing dom node logic. Here is what I'm seeing. 

* The node is updated on fetchChallengesCompleted, this only happens after the dom is  mounted.
* The cdu logic is triggered by react
* The updateMapScroll finds the node (since we are already mounted)
* The logic to updateThe map scroll does its magic.

What I'm failing to see the map scroll down to the challenge element.


I believe it is this logic here: 

```js
if (itemBottom + offset < containerHeight) {
      // no need to scroll
      return;
    }
```
<img width="359" alt="rosetta_code_problems__comma_quibbling___freecodecamp" src="https://user-images.githubusercontent.com/6775919/35883801-67a5d234-0b3d-11e8-9ba0-d32013024ae5.png">

That logic always catches and ends the function call without any further logic which I don't think was the intention. It seems to be due to the map panels not updating until after this logic is triggered. Initially all the challenge nodes are hidden but within the container window, the logic never continues past this point

Reply:
> What I'm failing to see the map scroll down to the challenge element.

There was an error in css (forget to rename some props during copy/paste from component).
I've fixed it with new commit. So everything have to be fine now.

As for the checks, that one will prevent scrollUpdate during the first run:
`if ( containerHeight === containerScrollHeight ) {`

It happens IMO because browser needs some time to update map scrollHeigh prop when new nodes (challenges) rendered.

I used timeouted call of updateMapScroll first (without retries) and it works fine with delay of 300ms. If delay is < 300 sometimes scroll is not updated. That is why I've implemented retry logic instead of fixed tiemout.